File: LoginDemo/LoginDemo/ViewModel/Common/NotificationObject.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using System.Linq;

namespace LoginDemo.ViewModel.Common
{
    public abstract class NotificationObject : INotifyPropertyChanged, IDataErrorInfo
    {
        #region 属性修改通知

        public event PropertyChangedEventHandler PropertyChanged;

        /// <summary>
        /// 发起通知
        /// </summary>
        /// <param name="propertyName">属性名</param>
        public void RaisePropertyChanged(string propertyName)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }

        #endregion

        #region 数据验证

        public string Error
        {
            get { return ""; }
        }

        public string this[string columnName]
        {
            get
            {
                var vc = new ValidationContext(this, null, null);
                vc.MemberName = columnName;
                var res = new List<ValidationResult>();
                var result = Validator.TryValidateProperty(this.GetType().GetProperty(columnName).GetValue(this, null), vc, res);
                if (res.Count > 0)
                {
                    return string.Join(Environment.NewLine, res.Select(r => r.ErrorMessage).ToArray());
                }
                return string.Empty;
            }
        }

        /// <summary>
        /// 页面中是否所有控制数据验证正确
        /// </summary>
        public virtual bool IsValid { get; set; }

        #endregion
    }
}

File: LoginDemo/LoginDemo/ViewModel/Common/NotificationObject.vb
Imports System.ComponentModel
Imports System.ComponentModel.DataAnnotations

Public MustInherit Class NotificationObject
    Implements INotifyPropertyChanged, IDataErrorInfo
#Region "属性修改通知"

    Private Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged
    ''' <summary>
    ''' 发起通知
    ''' </summary>
    ''' <param name="propertyName">属性名</param>
    Public Sub RaisePropertyChanged(ByVal propertyName As String)
        RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
    End Sub
#End Region
#Region "数据验证"

    Public ReadOnly Property [Error]() As String Implements IDataErrorInfo.Error
        Get
            Return ""
        End Get
    End Property

    Default Public ReadOnly Property Item(columnName As String) As String Implements IDataErrorInfo.Item
        Get
            Dim vc = New ValidationContext(Me, Nothing, Nothing)
            vc.MemberName = columnName
            Dim res = New List(Of ValidationResult)()
            Dim result = Validator.TryValidateProperty(Me.[GetType]().GetProperty(columnName).GetValue(Me, Nothing), vc, res)
            If res.Count > 0 Then
                Return String.Join(Environment.NewLine, res.[Select](Function(r) r.ErrorMessage).ToArray())
            End If
            Return String.Empty
        End Get
    End Property

    ''' <summary>
    ''' 页面中是否所有控制数据验证正确
    ''' </summary>
    Public Overridable Property IsValid() As Boolean
        Get
            Return m_IsValid
        End Get
        Set
            m_IsValid = Value
        End Set
    End Property
    Private m_IsValid As Boolean


#End Region
End Class